src/sbvr-api/sbvr-utils.ts

```diff
@@ -372,12 +372,60 @@ export const isModelNew = async (
 	return !cachedIsModelNew.has(modelName);
 };
 
+const bindsForAffectedIds = (
+	bindings: AbstractSQLCompiler.Binding[],
+	request?: Pick<
+		uriParser.ODataRequest,
+		| 'vocabulary'
+		| 'abstractSqlModel'
+		| 'method'
+		| 'resourceName'
+		| 'affectedIds'
+	>,
+) => {
```

Comment:
```suggestion
): ODataBinds => {
```

Reply:
@Page- `ODataBinds` doesn't actually cover the case needed for this particular function, which is map and not an array: https://github.com/balena-io/pinejs/pull/601/files#diff-824816fb30b8d1b6c2ec737540770325de0f51a1d9d8fca9e44bedcae76007d0R397. Even though this form is supported just fine: https://github.com/balena-io/pinejs/blob/master/src/sbvr-api/abstract-sql.ts#L117-L128.

IIRC the problem here is that `bindsForAffectedIds` can't know which index are the ones used by the narrowing snippet so that ID is bound by name as supported by the asql compiler and pine.

`ODataBinds` is defined by `odata-parser` so maybe it's worth expanding the type definition there first?

Reply:
@Ereski `ODataBinds` is defined as
```ts
export interface ODataBinds extends Array<any> {
	[key: string]: any;
}
```
which is both an array and a dictionary object so it should be fine

Reply:
@Page- It's not because it requires the usual symbols present in an array `extends Array`. Can't take an existing array and add extra entries either because of [isArray](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/isArray). Would have to be something closer to:

```ts
export type ODataBinds =
  Array<any>
  | {
    [key: string]: any;
  }
```

But what would the ramifications be?

Reply:
@Ereski it's specifically both an array and an object - where is the `isArray` check that's causing you problems if you go that route?

Reply:
@Page- It's an array with arbitrary extra fields. But in this case in particular it should be *either* an array or an object, not both. I'm going to explain the whole logic here to show the typing issue better and as a sanity check to myself too.

In `getAndCheckBindValues`, values to be bound are passed through `request.odataBinds`, which has type `OdataBinds` and is what you posted:

```ts
export interface ODataBinds extends Array<any> {
	[key: string]: any;
}
```

`getAndCheckBindValues` retrieves the actual value for each binding in the rule's SQL from `request.odataBinds`, and those values are used when running the rule: https://github.com/balena-io/pinejs/blob/7352cf516f188b21485647fb944faa649ee9c081/src/sbvr-api/sbvr-utils.ts#L442-L451. It additionally checks if those values are valid.

`getAndCheckBindValues` accepts multiple shapes for `odataBinds` because there is more than one way to name a binding: https://github.com/balena-io-modules/abstract-sql-compiler/blob/master/src/AbstractSQLCompiler.ts#L141. Where

- `['Bind', string, string]` defines a reference to a qualified table.column if I'm reading this correctly: https://github.com/balena-io/pinejs/blob/master/src/sbvr-api/abstract-sql.ts#L84-L105
- `['Bind', number]` defines a positional binding and `request.odataBinds` must be an array: https://github.com/balena-io/pinejs/blob/master/src/sbvr-api/abstract-sql.ts#L106-L116
- `['Bind', string]` defines a named binding and `request.odataBinds` must be an object: https://github.com/balena-io/pinejs/blob/master/src/sbvr-api/abstract-sql.ts#L117-L127

I made an error in my previous comment and `isArray` doesn't matter here, but the problem is the same: an array is not an arbitrary object.

In the case of narrowing, bindings fall in the last category above, using named bindings. The name here being the name of the table which that binding narrows. In this case, `request.odataBinds` *must not* be an array for one reason: an `Array<any>` implies predefined symbols with predefined behaviors and predefined types. If, by coincidence, the name of a table clashes (and possibly overwrites) one of those symbols then the `Array` contract is broken.

In this particular setup that's not a fatal problem for almost all cases because all `getAndCheckBindValues` does is check the array's size and index into it. But it would still be a bug and a potential problem in the future. The only exception AFAIK would be if the name of a table was `length`, which would cause an exception.

Having `odataBinds` be a pure map avoids that `length` and similar errors, and redefining `ODataBinds` to be either an array or a map (instead of a possibly mangled array) resolves everything given the appropriate type checks.